### PR TITLE
Added Filters

### PR DESCRIPTION
**Pull request by Peternator7**
Implemented Several of the filters requested on this-week-in-rust. First, Last, and Join for the Array filters. Rust fmt also added a comma here and there.

### CODE REVIEW

**Comment by Keats** [comment]
Thanks for that!
I think `first` and `last` should not be errors as it seems common enough that it can happen and that isn't necessarily an error. Returning `""` should be enough and maybe `None` later on when Tera will have its own serialization.

File: src/filters/array.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    if let Some(val) = arr.last() {
+        Ok(val.clone())
+    } else {
+        Err(TeraError::EmptyArray("last".to_string()))
+    }
+}
+
+pub fn join(value: Value, args: HashMap<String, Value>) -> TeraResult<Value> {
+    let arr = try_get_value!("join","value",Vec<Value>, value);
+    let separator = match args.get("d") {
```
Comment (Keats) [inline src/filters/array.rs]: can you name that arg `sep`?

File: src/filters/array.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        Some(val) => try_get_value!("truncate", "d", String, val.clone()),
+        None => "".to_string(),
+    };
+    let mut reduction = arr.iter()
+        .fold("".to_string(),
+              |acc, val| acc + &*separator + &*val.render());
+    if reduction.len() > 0 {
+        // if we had any elements, trim the separator off the front.
+        reduction = reduction.split_at(separator.len()).1.to_string();
+    }
```
Comment (Keats) [inline src/filters/array.rs]: `Vec` have a join method https://doc.rust-lang.org/std/slice/trait.SliceConcatExt.html#tymethod.join, is it possible to use it?

File: src/filters/array.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    } else {
+        Err(TeraError::EmptyArray("last".to_string()))
+    }
+}
+
+pub fn join(value: Value, args: HashMap<String, Value>) -> TeraResult<Value> {
+    let arr = try_get_value!("join","value",Vec<Value>, value);
+    let separator = match args.get("d") {
+        Some(val) => try_get_value!("truncate", "d", String, val.clone()),
+        None => "".to_string(),
```
Comment (Keats) [inline src/filters/array.rs]: default to `" "` would be better imo

Comment (Keats) [inline src/filters/array.rs]: maybe not actually, leave it like that for now

Comment (Peternator7) [inline src/filters/array.rs]: Yeah, I thought about that. The docs default to "" though so I thought at least initially consistency might be best.

File: src/errors.rs [hunk 4]
```diff
@@ -28,7 +28,7 @@ quick_error! {
             description("filter not found")
         }
 
-        // Runtime errors
+// Runtime errors
```
Comment (Keats) [inline src/errors.rs]: can you align that again?

**Comment by Keats** [comment]
Thanks for that!

**Pull request merged**